vault/generate_root.go

```diff
@@ -263,34 +264,67 @@ func (c *Core) GenerateRootUpdate(ctx context.Context, key []byte, nonce string,
 		}, nil
 	}
 
-	// Recover the master key
-	var masterKey []byte
+	// Combine the key parts
+	var combinedKey []byte
 	if config.SecretThreshold == 1 {
-		masterKey = c.generateRootProgress[0]
+		combinedKey = c.generateRootProgress[0]
 		c.generateRootProgress = nil
 	} else {
-		masterKey, err = shamir.Combine(c.generateRootProgress)
+		combinedKey, err = shamir.Combine(c.generateRootProgress)
 		c.generateRootProgress = nil
 		if err != nil {
 			return nil, errwrap.Wrapf("failed to compute master key: {{err}}", err)
 		}
 	}
 
 	switch {
-	case c.recoveryMode:
-		if err := c.barrier.Unseal(ctx, masterKey); err != nil {
-			c.logger.Error("root generation aborted, recovery token verification failed", "error", err)
+	case c.seal.RecoveryKeySupported():
+		// Ensure that the combined recovery key is valid
+		if err := c.seal.VerifyRecoveryKey(ctx, combinedKey); err != nil {
+			c.logger.Error("root generation aborted, recovery key verification failed", "error", err)
 			return nil, err
 		}
+		// If we are in recovery mode, then retrieve
+		// the stored keys and unseal the barrier
+		if c.recoveryMode {
+			if !c.seal.StoredKeysSupported() {
```

Comment:
Why not just call Core.UnsealWithStoredKeys?

Reply:
Because we don't want to actually do an unseal and are interested in only unsealing the barrier. Making an actual unseal will trigger `unsealInternal` which does more things than necessary.

Reply:
D'oh, good point.